### PR TITLE
Release pipeline for C++ driver

### PR DESCRIPTION
**Pull request by krishnangovindraj**
## Usage and product changes
Introduce build targets & jobs for the release pipeline of the C++ driver

## Implementation
* Introduce assemble & deploy & assembly test targets for all supported platforms the C++ driver in `cpp/BUILD`
* Update circleci config to deploy snapshots & releases on appropriate branches.
* Introduces the C++ driver readme

### CODE REVIEW

File: cpp/BUILD [hunk 1]
```diff
@@ -177,6 +177,17 @@ deploy_artifact(
...
+    name = "assemble-cpp-driver",
+    actual = select({
+        "@vaticle_bazel_distribution//platform:is_linux_arm64" : ":assemble-linux-arm64-targz",
+        "@vaticle_bazel_distribution//platform:is_linux_x86_64" : ":assemble-linux-x86_64-targz",
+        "@vaticle_bazel_distribution//platform:is_mac_arm64" : ":assemble-mac-arm64-zip",
+        "@vaticle_bazel_distribution//platform:is_mac_x86_64" : ":assemble-mac-x86_64-zip",
+        "@vaticle_bazel_distribution//platform:is_windows_x86_64" : ":assemble-windows-x86_64-zip"
+    })
+)
+
```
Comment (krishnangovindraj) [inline cpp/BUILD:190]: Convenient single target that switches platform. Not really needed.

File: cpp/lib/concept/value/value.cpp [hunk 2]
```diff
@@ -32,19 +32,18 @@
 namespace TypeDB {
 
 std::string Value::formatDateTime(DateTime t) {
-    std::time_t epochSeconds = std::chrono::time_point_cast<std::chrono::seconds>(t).time_since_epoch().count();
-    std::stringstream ss;
-    ss << std::put_time(std::localtime(&epochSeconds), "%FT%T");
-    return ss.str();
```
Comment (krishnangovindraj) [inline cpp/lib/concept/value/value.cpp:38]: This worked fine, but I've changed it to reflect parseDateTime

File: cpp/lib/concept/value/value.cpp [hunk 3]
```diff
@@ -32,19 +32,18 @@
...
+    char buf[32];
+    strftime(buf, 32, "%Y-%m-%dT%H:%M:%S", std::gmtime(&epochSeconds));
+    return std::string(buf);
 }
 
 DateTime Value::parseDateTime(const std::string& s) {
     std::tm tm = {};
-    std::stringstream ss(s);
     if (s.find("T") != std::string::npos) {
-        ss >> std::get_time(&tm, "%FT%T");
```
Comment (krishnangovindraj) [inline cpp/lib/concept/value/value.cpp:45]: get_time broke on linux/gcc.
strptime isn't defined on windows.

File: cpp/test/assembly/test.cpp [hunk 4]
```diff
@@ -24,14 +24,19 @@
...
+        if (driver.databases.contains(dbName)) {
+            driver.databases.get("testAssembly").drop();
+        }
+        if (driver.databases.contains(dbName)) return 1;
+        driver.databases.create(dbName);
+        if (!driver.databases.contains(dbName)) return 1;
+        std::cout << "Success!" << std::endl;
+    } catch (TypeDB::DriverException e ) {
+        std::cout << "Caught TypeDB::DriverException: " << e.code() << "\n" << e.message()  << std::endl;
+        return 2;
```
Comment (krishnangovindraj) [inline cpp/test/assembly/test.cpp:39]: Catching this exception give us a helpful message, so we know it's not the assembly that failed but something earlier.

File: cpp/test/behaviour/tests.bzl [hunk 5]
```diff
@@ -26,20 +26,20 @@ def typedb_behaviour_cpp_test(
...
+        **kwargs,
     )
     
     native.cc_test(
         name = name + "-cloud",
         deps = ["//cpp/test/behaviour/steps:cloud-steps"] + deps,
         data = [feature, "//tool/test/resources:certificates"] + data,
         args = ["$(location " + feature + ")"] + args,
-        env = {"ROOT_CA": "tool/test/resources/encryption/ext-grpc-root-ca.pem"} | env,
+        **kwargs,
```
Comment (krishnangovindraj) [inline cpp/test/behaviour/tests.bzl:44]: We seem to pass through the CLI, so we avoid hardcoding it here.
Had forgotten the `**kwargs`

**Review by krishnangovindraj — COMMENTED** [review]
self-review

File: .circleci/config.yml [hunk 6]
```diff
@@ -286,6 +286,170 @@ commands:
...
+    parameters:
+      assembly:
+        type: string
+    steps:
+      - run: |
+          ulimit -n 100000
+          bazel build //c:assemble-clib-driver
+          export ASSEMBLY=<<parameters.assembly>>
+          export ASSEMBLY_BASE=$(echo $ASSEMBLY | cut -f 1 -d '.')
+          export OSX_ARCH=$(echo $ASSEMBLY_BASE | cut -f 5 -d '-')
```
Comment (flyingsilverfin) [inline .circleci/config.yml:330]: Is this for all `unix` or just `mac`?

File: .circleci/config.yml [hunk 7]
```diff
@@ -286,6 +286,174 @@ commands:
...
+          mkdir -p test_assembly_clib
+          tar -xf bazel-bin/c/$ASSEMBLY --directory test_assembly_clib
+          ( cd test_assembly_clib && cmake ../c/tests/assembly -DTYPEDB_ASSEMBLY=$(pwd)/$ASSEMBLY_BASE -DCMAKE_OSX_ARCHITECTURES=$OSX_ARCH && cmake --build . --config release )
+          
+          tool/test/start-core-server.sh
+          sleep 3
+          export TYPEDB_LIB_PATH=$ASSEMBLY_BASE/lib
+          ( cd test_assembly_clib && DYLD_LIBRARY_PATH=$TYPEDB_LIB_PATH LD_LIBRARY_PATH=$TYPEDB_LIB_PATH ./test_assembly && export TEST_SUCCESS=0 || export TEST_SUCCESS=1)
```
[the remark this answers is not in the document]
Comment (flyingsilverfin) [reply .circleci/config.yml:338]: This makes me ask the same question

File: c/BUILD [hunk 8]
```diff
@@ -50,6 +58,163 @@ rust_cbindgen(
...
+        "//rust:typedb_driver_sync",
+
+        "@vaticle_typeql//rust:typeql",
+
+        "@crates//:chrono",
+        "@crates//:itertools",
+        "@crates//:env_logger",
+        "@crates//:log",
+    ],
+)
```
[the remark this answers is not in the document]
Comment (flyingsilverfin) [reply c/BUILD:79]: We could at least variabilise the `deps` into a list and use them in both for now

**Review by flyingsilverfin — APPROVED** [review]

**Pull request merged**